infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/transport/sample/GRPCLogReplicationServerChannelAdapter.java

```diff
@@ -59,13 +63,30 @@ public void send(CorfuMessage.RequestMsg msg) {
     public CompletableFuture<Boolean> start() {
         CompletableFuture<Boolean> serverCompletable = new CompletableFuture<>();
         try {
-            server.start();
-            serverCompletable.complete(true);
-            log.info("Server started, listening on {}", port);
-        } catch (Exception e) {
-            log.error("Caught exception while starting server on port {}", port, e);
-            serverCompletable.completeExceptionally(e);
+            IRetry.build(IntervalRetry.class, () -> {
+                try {
+                    server.start();
+                    serverCompletable.complete(true);
+                    log.info("Server started, listening on {}", port);
+                } catch (IllegalStateException ise) {
+                    // If a gRPC server has started or shutdown, the same instance cannot be used for
+                    // starting a gRPC server again, and a new instance has to be created. This is a
+                    // limitation introduced by gRPC.
+                    log.warn("gRPC server already started or shutdown, instantiating a new one.", ise);
+                    this.server = ServerBuilder.forPort(port).addService(service)
```

Comment:
it may be good to move this creation to a common method which can be used from here and the constructor so that any future changes in this part are reflected for both workflows

Reply:
yes I agree, extracted a helper method for creating server instance